### ISSUE TITLE
[TimeSeries] Potential bug in quantile predictions

### ISSUE THREAD

**Issue by Innixma** [autogluon/autogluon#4062]
Users noticed some strange behavior where lower quantiles would predict constant values while other quantiles would predict time-varying values. 

When they manually hacked a work-around, they achieved significantly better scores. Something to consider looking into:

https://www.kaggle.com/competitions/probabilistic-forecasting-i-temperature/discussion/491759#2739646

https://www.kaggle.com/code/samvelkoch/autogluon-ts-experiment-1-6614/comments#2740055

Note that the issue appears to exist in mainline:
> I've checked it on 'night build' (pip install --pre) and got same result

**Comment by samvelkoch** [comment]
Thanks for open an issue here Nick. I'll let you know here if will find something else. 

I tried already to predict only 0.025, 0.05, 0.1 and 0.15 quantiles only form that data and achieved same flat shape. Will investigate issue next days.

**Comment by canerturkmen** [comment]
Hi @samvelkoch ,

Thank you so much for giving AutoGluon-TimeSeries a try 📈 !!

I took a closer look at the issue. While it's not a bug, it is an artifact of multiple things happening at once. 

1/ First, in this run all deep learning models appear to be failing due to a torch dependency issue. 2/ Even without this issue, however, these models would be essentially impossible to fit well with a forecast horizon of 5K time steps. Due to these, AG is falling back to two models which it builds a weighted ensemble with: DirectTabular and NPTS.

NPTS is a simple baseline which takes (weighted) samples from the time series values in the context. DirectTabular uses LightGBM from AG-Tabular and featurization from the MLForecast library to build a booster for each quantile. In this example, the boosted tree mainly relies on `feature_BB` and `feature_CA` which have lighter tails in the forecast horizon and do not affect the forecast. Hence the near-constant forecasts, mainly due to how DirectTabular independently fits different boosters for different quantiles (forecasts shown below).

Another issue is the use of `refit_every_n_windows`. Setting this parameter and not setting `num_val_windows` leads to `num_val_windows` defaulting to 1. Therefore, models are backtested for only a single window and `refit_every_n_windows` does not take effect. I realize this was not mentioned clearly in the documentation and made a note to fix this going forward. Thank you for bringing this to our attention.

I also shared a quick competition [notebook](https://www.kaggle.com/code/canerturkmen/pfi-autogluon-timeseries-medium-quality), downsampling the data for a quick run with AutoGluon and it also seems to be doing fairly well (1.73). I hope this helps.

I will now close the issue but feel free to comment to keep the discussion going. Thanks!

![output](https://github.com/autogluon/autogluon/assets/1311949/6f94283f-886f-42f7-8bb0-d92288df62c0)

**Comment by samvelkoch** [comment]
Thank you @canerturkmen for investigation. Looking forward to try new AG version and your findings in this competition

**Comment by Innixma** [comment]
@canerturkmen I'm curious why the model consistently chooses to predict a constant value for not just one but multiple of the lower quantiles. Is there something in our validation logic that is incentivizing models to do strange things in these lower quantiles? Why is the same not happening for the upper quantiles? Is it a quirk of the evaluation metric?

**Comment by Innixma** [comment]
Because @samvelkoch observed improvement when adding a hack to work-around this, I wonder if this issue goes a bit deeper than one might see at first glance.

**Comment by valeman** [comment]
@canerturkmen how does AutoGluonTS aggregate quantile prediction from NPTS and DirectTabular? There are two potential issues here 1) from what I understand NPTS does not have quantile as direct input (my understanding could be wrong), is there some trick AutoGluonTS uses to let NPTS know prediction interval confidence? 2) How is ensemble of quantiles built - would be good to understand if there is some reference to it in documentation or some paper please.

**Comment by canerturkmen** [comment]
Thanks for your comment @valeman !

The AG-TS [paper](https://assets.amazon.science/24/de/d7daa2114569bb9c5c4feefd967e/autogluon-timeseries-automl-for-probabilistic-time-series-forecasting.pdf) goes into slightly more detail. To be specific, the aggregation is a simple weighted average of quantiles from models, the "coarse-grained" approach [here](https://www.jmlr.org/papers/volume24/22-0799/22-0799.pdf), and the weights are learned in a hill-climbing fashion using the "ensemble selection" method outlined by [Caruana et al](https://www.cs.cornell.edu/~alexn/papers/shotgun.icml04.revised.rev2.pdf).

NPTS is about one of the most naive baselines you can imagine, but it has served well in some use cases so far. One just takes samples from the context window and uses those to build Monte Carlo estimates of quantiles in the prediction window. It is therefore "non-parametric" and unaware of the target quantile levels.

**Comment by canerturkmen** [comment]
@Innixma thanks for keeping the discussion going. about the "hack," I would expect post-hoc calibration with or without a hack to improve greatly in this competition as it is primarily about calibrating a single time series forecast well across a very long forecast horizon. This is quite an edge case for AG-TS. I think AG-TS is operating as intended by falling back to TabularPredictor. I will check after the release why TabularPredictor is churning out constant quantiles.

**Comment by canerturkmen** [comment]
btw, another issue appears to be that deep learning models in AG-TS are not working on the Kaggle container image. I'm currently looking into this, tracking here: #4090

**Comment by valeman** [comment]
That’s great thank you for these details @canerturkmen

**Comment by valeman** [comment]
@canerturkmen one more question please, in this Amazon paper Deep Non-Parametric Time Series Forecaster Deep NPTS does seem to be producing prediction intervals based on user specific level (Figure 1). https://arxiv.org/abs/2312.14657 How were these produces, is there some example of code showing which parameter a user can specify in NPTS or Deep NPTS (or both) to control for prediction interval confidence level. This could be a very useful non parametric benchmark and also showcase of these Amazon methods for the Kaggle competition to promote best practices of having probabilistic benchmark similar to what is done in point forecasting.

**Comment by canerturkmen** [comment]
Sure @valeman !

NPTS simply boils down to taking samples from the context window and constructing forecasts based on these, across the forecast horizon. The samples, in turn, are taken w.r.t. a simple distribution such as uniform or geometric. This is the approach we implement here. DeepNPTS, as far as I am knowledgeable, learns the weights (from which the predictions will be sampled) via a neural network, across multiple items in the data set. As such, the model can 'learn' patterns like seasonality or trends for which parts of the context to sample from. We do not have DeepNPTS provided here, although I think the reference implementation is [here](https://github.com/awslabs/gluonts/blob/dev/src/gluonts/torch/model/deep_npts/_estimator.py).

In either case the quantiles themselves are not user-specified. Rather, the user specifies a `number_of_samples` parameter to control the sampling error around the quantile estimates. In our interface, that is [here](https://github.com/autogluon/autogluon/blob/master/timeseries/src/autogluon/timeseries/models/local/npts.py#L25) in the code and [here](https://auto.gluon.ai/stable/tutorials/timeseries/forecasting-model-zoo.html#autogluon.timeseries.models.NPTSModel) in the docs.

About a benchmark for probabilistic forecasting: that sounds like something that we all definitely need! In our experience, most methods used today do tend to undercover, and NPTS's nonparametric nature helps counter this especially in pathologically right skewed cases such as lumpy demand. However, we believe both forecast calibration and calibrated forecasts in forecast aggregation are significant open questions we're also keeping an eye on.

**Comment by valeman** [comment]
Thanks you @canerturkmen this is very useful, I will need to deep dive into code to learn more about the relationship between number of samples and PI confidence levels.
